Bug/ai move computation

NeuralCoder3 opened the pull request:
This PR fixes the threaded computation of the AI moves.
By doing so, a bug in the board constructor is fixed.
`Arrays.copy` creates a (deep) copy of the array by cloning each element.
However, the Marks are copied by reference.
The new board references the original Marks.
This leads to errors when `setMarkAt` changes symbols of Marks.

As mentioned in the discourse discussion, the GUI does not freeze.
It handles the moves correctly (at least on my system).
The problem was that
* the move of the AI was never played resulting in no visible change
* the board was silently mutated which together with the silently failing `play` method resulting in no visible change

Additionally, the `play` method was changed to be in compliance with defensive programming.
From the name and semantics, it is expected that this method should play a move.
A silent failure when the move is not possible can be confusing and hides errors (as the one above).
In the case here, the AI would try to play a move on a (silently) changed board and would do nothing as the AI computation has filled the board.
This method should only be called with correct moves. The caller should handle invalid cases or catch errors.

The main fix for the threading problem is to start the computation using a worker thread (SwingWorker does this) and pass the result back to the EDT for taking action (in the `done` method).

NeuralCoder3 committed:
deep copy constructor for Board to avoid aliasing

NeuralCoder3 committed:
direct error handling according to defensive programming

NeuralCoder3 committed:
threaded AI move computation

NiGoe04 commented:
> This PR fixes the threaded computation of the AI moves. By doing so, a bug in the board constructor is fixed. `Arrays.copy` creates a (deep) copy of the array by cloning each element. However, the Marks are copied by reference. The new board references the original Marks. This leads to errors when `setMarkAt` changes symbols of Marks.
> 
> As mentioned in the discourse discussion, the GUI does not freeze. It handles the moves correctly (at least on my system). The problem was that
> 
> * the move of the AI was never played resulting in no visible change
> * the board was silently mutated which together with the silently failing `play` method resulting in no visible change
> 
> Additionally, the `play` method was changed to be in compliance with defensive programming. From the name and semantics, it is expected that this method should play a move. A silent failure when the move is not possible can be confusing and hides errors (as the one above). In the case here, the AI would try to play a move on a (silently) changed board and would do nothing as the AI computation has filled the board. This method should only be called with correct moves. The caller should handle invalid cases or catch errors.
> 
> The main fix for the threading problem is to start the computation using a worker thread (SwingWorker does this) and pass the result back to the EDT for taking action (in the `done` method).

NiGoe04 commented:
I did consider the case where the move is not valid but thought, that in this case the player could just do the next move manually, since play() would return and onMoveSelected would run again with the player move once a button is clicked. Or is this not the case? To your change: 
        if (!board.moveIsPossible(move.index())) return;
        if (!board.moveIsPossible(move.index())) {
            throw new IllegalArgumentException("Move is not possible, index: " + move.index()+ " is already "+ board.getMarks()[move.index()].symbol());
        }
Wouldnt the IllegalArgumentException never be thrown, because if the condition is truem the function would have returned by the line before?
More interesting is the deep copy of the board, thanks a lot for pointing this out i wouldnt have noticed this.

NeuralCoder3 commented:
> Or is this not the case? 

Yes, this would be possible. 
However, it would not be intuitive to me to handle this in the play method.
Usually, it is a good idea to modularize the application.
In this case, it might be easier and more readable to have a check when a player selects the action and tell it before even attempting to play it.
This coincides with a common model called Model-View-Controller where you want to keep UI, handling/communication, and logic as separated as possible to increase modularity.

I think you read this wrong.
You can take a look at the [Files Changed](https://github.com/NiGoe04/TicTacToeAi/pull/1/files#diff-14428398adb6fdd21ed6bf7b5fcb4b617e2cd1008465446ef39b178454452353R67) tab (and even comment in there):
![2023-08-21_22-21-36](https://github.com/NiGoe04/TicTacToeAi/assets/43005464/fb9c06cb-a59d-4917-a217-30aa58c219aa)

I removed the `return` line and replaced it with the `IllegalArgumentException` (although one could argue that a custom exception might be more appropriate here).

> More interesting is the deep copy of the board, thanks a lot for pointing this out i wouldnt have noticed this.

This is the main reason I changed the function to the defensive programming style.
Without it, the bug(s) are hard to detect and go unnoticed or seem to originate at a different location.

NiGoe04 commented:
Ah yes you are right, sometimes my red and green color blindness gets in my way ;)
Thanks for the help on the project, i will now try to figure out why my AI plays kinda random moves.

NeuralCoder3 commented:
> Ah yes you are right, sometimes my red and green color blindness gets in my way ;)

Github supports multiple color schemes. You can change it [here](https://github.com/settings/appearance).
Light/Dark Protanopia & Deuteranopia should probably help.
